projects/components/src/filtering/filter-bar/filter-chip/filter-chip.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable } from '@angular/core';
+import { FilterBuilderLookupService } from '../../filter/builder/filter-builder-lookup.service';
+import { AbstractFilterBuilder } from '../../filter/builder/types/abstract-filter-builder';
+import { IncompleteFilter } from '../../filter/filter';
+import { FilterAttribute } from '../../filter/filter-attribute';
+import { FilterOperator } from '../../filter/filter-operators';
+import { FilterParserLookupService } from '../../filter/parser/filter-parser-lookup.service';
+import { SplitFilter } from '../../filter/parser/parsed-filter';
+import { AbstractFilterParser, splitFilterStringByOperator } from '../../filter/parser/types/abstract-filter-parser';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FilterChipService {
+  public constructor(
+    private readonly filterBuilderLookupService: FilterBuilderLookupService,
+    private readonly filterParserLookupService: FilterParserLookupService
+  ) {}
+
+  public autocompleteFilters(attributes: FilterAttribute[], text: string = ''): IncompleteFilter[] {
+    return attributes
+      .filter(attribute => this.filterBuilderLookupService.isBuildableType(attribute.type))
+      .flatMap(attribute => this.autocompleteFilter(attribute, text))
```

Comment:
I asked this in the other PR but don't recall the answer and have the same question again - if we're passing `text` into `autocompleteFilter()`, why would it ever return filters that don't match `text`?

Reply:
The returned filters should always have a `userString` that _includes_ the user provided `text`. Additionally, once the entire string matches and the user hits `space`, they are also provided an option for each operator available to that attribute.

Reply:
> The returned filters should always have a userString that includes the user provided text.

In other words, `isStringMatch` as defined on line 28 will always be true, right? 
```typescript
const isStringMatch = incompleteFilter.userString.toLowerCase().includes(text.trim().toLowerCase());
```
And if that's always true, then `filterMatchesUserText` will always return true and the filter is not needed.

Reply:
Since we are testing each attribute for a match, it won't always be true.

Reply:
OK I dug a couple calls down and I think I see it now, but I do still find it very convoluted - if the text does match, we use it to build an incomplete filter, else it always returns an incomplete filter representing that attribute, right?

My expectation as a reader would be that autocompleteFilter would only return something that matches.

Reply:
Correct. We turn each attribute into an IncompleteFilter, and then filter the ones that don't match back out.

Agreed its a little convoluted, but the reason we don't filter first is because we need to build the userString and filter against that.

Reply:
> the reason we don't filter first is because we need to build the userString and filter against that.

But doesn't that method build the userString as part of what it's doing? That's how it's able to build the incomplete filter in the first place

Reply:
Yep, that's what I mean. Ideally, we would filter first and then build the `IncompleteFilter`s, but the reason we don't swap the order of `autocompleteFilter()` and `filterMatchesUserText()` is because we need the `userString` from the `IncompleteFilter`s that `autocompleteFilter()` produces for the `filterMatchesUserText()` step.

Reply:
That makes sense why it can't be done before going into autocomplete. I was more focusing on the implementation of autocomplete - it returns an array of IncompleteFilters, so it could filter them before returning so that it only returns matches.

That would be my expectation as a reader, but if you disagree we can leave it.

Reply:
Ah, I see. That's a fair concern. I do like the way the current method chain reads, but maybe because I've been staring at it for so long.

I just renamed those methods to maybe clear things up a bit.